### PR TITLE
test: improve type checks

### PR DESCRIPTION
**Pull request by zacharygolba**
**WIP**

This is an experimental refactor with the goal of improving the type declarations within the project to help ensure `1.0` stability.

### CODE REVIEW

**Comment by mutewinter** [comment]
@zacharygolba I see you removed a lot of type declarations for return values. Is that because they are implicit?

**Comment by mutewinter** [comment]
@zacharygolba where does `Reflect` come from? I see it in use, but not imported.

**Comment by mutewinter** [comment]
I think this PR deserves a better name at this point, perhaps something like `add better relationship support and refactor type systems`.

**Comment by mutewinter** [comment]
@zacharygolba were you able to remove `moment` because you implemented everything using native `Date`s?

File: src/utils/set-type.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * function can still benifit from type inference and safety as long as the
+ * return value type declaration is 100% accurate.
+ *
+ * WARNING:
+ * This function should rarely be used as it requires a good understanding of
+ * the flow type system to ensure that the function this util wraps is still
+ * type safe.
+ *
+ * @private
+ */
```
Comment (mutewinter) [inline src/utils/set-type.js]: Nicely documented!

**Comment by mutewinter** [comment]
I read through everything. I can't say I understood every bit, but it does look solid. Well done, sir!

**Comment by harianus** [comment]
> @zacharygolba where does Reflect come from? I see it in use, but not imported.

Reflect is a built-in object that provides methods for interceptable JavaScript operations. The methods are the same as those of proxy handlers. Reflect is not a function object, so it's not constructible.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect

**Comment by zacharygolba** [comment]
> I see you removed a lot of type declarations for return values. Is that because they are implicit?

That's correct. Nuclide has a widget that displays the percentage of what is covered by flow within a file which makes it easy to see when you can/should get away with type inference.

screen shot 2016-07-31 at 10 36 17 pm

**Comment by zacharygolba** [comment]
> where does Reflect come from? I see it in use, but not imported.

@harianus is spot on. It is just a different way of performing dynamic property access, assignment, constructor calls, etc. It's kind of like the inverse of `Proxy`.

[Flow does not like dynamic property accessors on items other than Array or Object instances.](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVBjGBDAzrsABQCc4AHAU2IBcBPACTwAswBvVMMAO2wFsKAXGFzViASy4BzANzpOGOFxHEArhmpxiACjKkyuAJRsOYAJAB5AEYArCuoB0eXGMlct1JmNwAaMLvKGspwAvqihmIoiYEzMYAC83BRIJORUdIy4TFrsnDz8QgDkUHBwBWEGsqjAwGAAosSkxAC0UMQUFGAAJrR5Yhhg2BgYFPj2qDAU1Nx8HQkAShSwdtT2kpNaMZm+BXkUBRXo1WAAKkwdxTDwCBKSYAqdHV5gAG7YMGKdYABS2K8AyhhxGQppYVFNYIgwB5SAgCNguFUalRGmNdvFoswANo7GYFAC6siAA) Using `Reflect` is essentially a way to say "this code is dynamic, get out of my way".

There is even an eslint rule in use for Lux now [`prefer-reflect`](http://eslint.org/docs/rules/prefer-reflect).

*Side Note:*

I've found it pretty useful to have an explicit way of handling extremely dynamic code.

```javascript
// This sticks out like a sore thumb compared to obj[randomKey];
Reflect.get(obj, randomKey);
```

**Comment by zacharygolba** [comment]
> I think this PR deserves a better name at this point, perhaps something like add better relationship support and refactor type systems.

Good call!

**Comment by zacharygolba** [comment]
> were you able to remove `moment` because you implemented everything using native Dates?

That's correct. Oddly enough, I did a search in the src directory and found that `moment` was only used in one place ([generating the timestamp for migrations in the CLI](https://github.com/postlight/lux/blob/master/src/packages/cli/commands/generate.js#L63)). It seems like overkill to use `moment` for this one task. Most Date related code in Lux uses the ISO 8601 format exclusively which fortunately is easy with native JS dates.

**Comment by zacharygolba** [comment]
@mutewinter thanks for the review man. You definitely deserve an award of some sort for going through this massive PR! 🏆

**Pull request merged**